### PR TITLE
Sha3 and Shake

### PR DESCRIPTION
**Pull request by fliphil**
* Also adds support for large/ endpoint

### CODE REVIEW

**Comment by bfussell** [comment]
I guess we'll need some UT, but we can make a separate user story for that.  Does the old hash UT still work with no leaks ?

File: src/acvp.c [hunk 1]
```diff
@@ -1476,6 +1491,102 @@ static ACVP_RESULT acvp_parse_login(ACVP_CTX *ctx) {
...
+    val = json_value_init_object();
+    obj = json_value_get_object(val);
+
+    /* 
+     * Cut off the https://name:port/ prefix and /results suffix
+     */
+    strstr_s((char *)url, ACVP_ATTR_URL_MAX, "/acvp/v1", 8, &substr);
+    strcpy_s(snipped_url, ACVP_ATTR_URL_MAX, substr);
+    strstr_s(snipped_url, ACVP_ATTR_URL_MAX, "/results", 8, &substr);
+    if (!substr) return ACVP_INVALID_ARG;
```
Comment (bfussell) [inline src/acvp.c]: goto err ? wouldn't arr_val be leaked otherwise ?

File: src/acvp_hash.c [hunk 2]
```diff
@@ -183,6 +199,226 @@ static ACVP_RESULT acvp_hash_mct_tc(ACVP_CTX *ctx,
...
+    unsigned int range = max_xof_bytes - min_xof_bytes + 1;
+
+    /*
+     * Initial Outputlen = (floor(maxoutlen/8) )*8
+     */
+    xof_len = (max_xof_bits / 8) * 8;
+    /* Convert from bits to bytes */
+    stc->xof_len = (xof_len + 7) / 8;
+
+#if 0
```
Comment (bfussell) [inline src/acvp_hash.c]: Still needed or something we need to look at later ?

File: src/acvp_hash.c [hunk 3]
```diff
@@ -183,6 +199,226 @@ static ACVP_RESULT acvp_hash_mct_tc(ACVP_CTX *ctx,
...
+                ACVP_LOG_ERR("crypto module failed the operation");
+                rv = ACVP_CRYPTO_MODULE_FAIL;
+                goto end;
+            }
+
+            /* Get the right-most 16bits and convert to an integer */
+#ifdef WIN32
+            rightmost_out_bits = _byteswap_ushort(*(uint16_t *)(stc->md + stc->md_len - 2));
+#else
+# if ACVP_HOST_LITTLE_ENDIAN
```
Comment (bfussell) [inline src/acvp_hash.c]: Got to be a 16 bit byte swap libc API that is endian agnostic I would think. Maybe not that supports windows and Mac ?

Comment (fliphil) [reply src/acvp_hash.c]: Do you have any specific in mind? I think ntohs is also endian dependent

**Pull request merged**